### PR TITLE
Multiserver support in pbs_init.d

### PR DESCRIPTION
**Pull request by hirenvadalia**
#### Describe Bug or Feature

Currently, the PBS init script doesn't support the start/stop/status of multiple servers.

#### Describe Your Change

Changed init script and its dependent scripts to start/stop/status of multiple servers
(Assumption, all servers configured in PBS_SERVER_INSTANCES are assumed for local host, as we don't have support for multi-host multi-server support yet, once we have that, we have to change these scripts again)

#### Link to Design Doc

None

#### Attach Test and Valgrind Logs/Output

[after.txt](https://github.com/subhasisb/pbspro/files/4563567/after.txt)
Screen Shot 2020-05-01 at 6 09 06 PM

### CODE REVIEW

File: src/cmds/scripts/pbs_dataservice [hunk 1]
```diff
@@ -361,16 +361,20 @@ case "$1" in
...
+			${PBS_EXEC}/bin/qstat -Bf >/dev/null 2>&1
 			if [ $? -eq 0 ]; then
-				echo "PBS server is running. Cannot stop PBS Data Service now."
-				server_pid="`ps -ef | grep "${PBS_EXEC}/sbin/pbs_server.bi[n]" | awk '{print $2}'`"
-				if [ -z "${server_pid}" ]; then
-					server_host="`${PBS_EXEC}/bin/qstat -Bf | grep 'server_host' | awk '{print $NF}'`"
-					if [ ! -z "${server_host}" ]; then
-						echo "PBS server is running on host ${server_host}."
+				if [ "x${PBS_SERVER_INSTANCES}" != "x" ]; then
+					echo "PBS Servers is running. Cannot stop PBS Data Service now."
```
Comment (bremanandjk) [inline src/cmds/scripts/pbs_dataservice:367]: Change this msg as

"PBS Servers are running. Cannot stop PBS Data Service now."

File: src/cmds/scripts/pbs_habitat.in [hunk 2]
```diff
@@ -532,6 +532,16 @@ if [ -f "$PBS_HOME/pbs_version" ]; then
 	old_pbs_version=`cat $PBS_HOME/pbs_version`
 fi
 
+# If multiserver is configure then use first server
+# in the given list of server instances for this
+# script
+if [ "x${PBS_SERVER_INSTANCES}" != "x" ]; then
+	_first_svr=${PBS_SERVER_INSTANCES%%,*}
```
Comment (bremanandjk) [inline src/cmds/scripts/pbs_habitat.in:539]: For better code readability, change this variable to _first_svr_port.

File: src/cmds/scripts/pbs_init.d.in [hunk 3]
```diff
@@ -265,6 +245,114 @@ check_hostname() {
 	return 1
 }
 
+get_server_pid() {
+	local _lockfn="" _pid=""
```
Comment (bremanandjk) [inline src/cmds/scripts/pbs_init.d.in:249]: What is the use of this local var "_pid"?, I am not seeing its use in this function.

File: src/cmds/scripts/pbs_init.d.in [hunk 4]
```diff
@@ -265,6 +245,114 @@ check_hostname() {
...
+		PBS_BATCH_SERVICE_PORT=$1 ${PBS_EXEC}/sbin/pbs_server
+		ret_val=$?
+	else
+		${PBS_EXEC}/sbin/pbs_server
+		ret_val=$?
+	fi
+	if [ ${ret_val} -eq 0 ]; then
+		echo "PBS server${_portmsg}"
+		return 0
+	elif [ ${ret_val} -eq 4 ] ; then
```
Comment (bremanandjk) [inline src/cmds/scripts/pbs_init.d.in:285]: To be consistent, remove the space between the closing bracket and semicolon.

File: src/cmds/scripts/pbs_init.d.in [hunk 5]
```diff
@@ -265,6 +245,114 @@ check_hostname() {
...
+	return 1
+}
+
+start_local_servers() {
+	local _started_any=0 _svr _host _port
+
+	if [ "x${PBS_SERVER_INSTANCES}" == "x" ]; then
+		if ! check_server
+		then
+			if [ -f ${pbslibdir}/init.d/limits.pbs_server ] ; then
```
Comment (bremanandjk) [inline src/cmds/scripts/pbs_init.d.in:310]: To be consistent, remove the space between the closing bracket and semicolon. 
and also where ever applicable.

File: src/server/pbsd_main.c [hunk 6]
```diff
@@ -1043,10 +1043,10 @@ main(int argc, char **argv)
...
 	}
 
-	/* make sure no other server is running with this home directory */
-
-	(void)sprintf(lockfile, "%s/%s/server.lock", pbs_conf.pbs_home_path,
-		PBS_SVR_PRIVATE);
+	if (get_max_servers() > 1)
+		(void)sprintf(lockfile, "%s/%s/server_%d.lock", pbs_conf.pbs_home_path, PBS_SVR_PRIVATE, pbs_server_port_dis);
+	else
+		(void)sprintf(lockfile, "%s/%s/server.lock", pbs_conf.pbs_home_path, PBS_SVR_PRIVATE);
```
Comment (bremanandjk) [inline src/server/pbsd_main.c:1049]: Can we have a common lock file name? so that we can avoid this if, else block. 
Since for single-server mode also the combination of "server_port.lock" would be unique enough to lock. Just a thought.

Comment (hirenvadalia) [reply src/server/pbsd_main.c:1049]: Well that's was my first idea, but there is a drawback in doing that right now because "server.lock" is hardcoded in testing framework, which I can change, but that will cause little pain for our team. So let's discuss this offline.

Comment (bremanandjk) [reply src/server/pbsd_main.c:1049]: okay, sure, It's up to you.

File: src/cmds/scripts/pbs_dataservice [hunk 7]
```diff
@@ -361,16 +361,20 @@ case "$1" in
...
-				server_pid="`ps -ef | grep "${PBS_EXEC}/sbin/pbs_server.bi[n]" | awk '{print $2}'`"
-				if [ -z "${server_pid}" ]; then
-					server_host="`${PBS_EXEC}/bin/qstat -Bf | grep 'server_host' | awk '{print $NF}'`"
-					if [ ! -z "${server_host}" ]; then
-						echo "PBS server is running on host ${server_host}."
+				if [ "x${PBS_SERVER_INSTANCES}" != "x" ]; then
+					echo "PBS Servers are running. Cannot stop PBS Data Service now."
+				else
+					echo "PBS server is running. Cannot stop PBS Data Service now."
+					server_pid="`ps -ef | grep "${PBS_EXEC}/sbin/pbs_server.bi[n]" | awk '{print $2}'`"
```
Comment (subhasisb) [inline src/cmds/scripts/pbs_dataservice:370]: do you really wanna check with the whole path? If somebody started manually that may not match the grep. I suggest just grepping for pbs_server.bi[n]

File: src/cmds/scripts/pbs_habitat.in [hunk 8]
```diff
@@ -532,6 +532,16 @@ if [ -f "$PBS_HOME/pbs_version" ]; then
 	old_pbs_version=`cat $PBS_HOME/pbs_version`
 fi
 
+# If multiserver is configure then use first server
+# in the given list of server instances for this
+# script
+if [ "x${PBS_SERVER_INSTANCES}" != "x" ]; then
```
Comment (subhasisb) [inline src/cmds/scripts/pbs_habitat.in:538]: WHy do you need to find and set the PBS_BATCH_SERVICE_PORT variable? Commands can automatically deduce this from the PBS_SERVER_INSTANCES. In fact, i think we should disallow setting both PBS_BATCH_SERVICE_PORT and PBS_SERVER_INSTANCS in the same pbs.conf file

File: src/cmds/scripts/pbs_init.d.in [hunk 9]
```diff
@@ -417,96 +505,111 @@ start_pbs() {
...
+		if check_server
+		then
+			_pid=`get_server_pid`
+			if [ "${my_hostname}" = "${_act_svr}" ]; then
+				if [ -z "${PBS_SECONDARY}" -o ${is_secondary} -eq 1 ]; then
+					echo "Shutting server down with qterm."
+				else
+					echo "This is active server, shutting down with qterm, secondary will take over."
+				fi
+				if [ -r "${PBS_HOME}/server_priv/qmgr_shutdown" ]; then
```
Comment (subhasisb) [inline src/cmds/scripts/pbs_init.d.in:536]: I dont think this method of shutdown still works. I have never seen a code about qmgr_shutdown in pbs_server. Plaese check if we can remove this?

File: src/cmds/scripts/pbs_init.d.in [hunk 10]
```diff
@@ -417,96 +505,111 @@ start_pbs() {
...
+				kill ${_pid}
+				echo "PBS server - was pid: ${_pid}"
+			fi
+		fi
+		${PBS_EXEC}/sbin/pbs_dataservice status > /dev/null 2>&1
+		if [ $? -eq 0 ]; then
+			${PBS_EXEC}/sbin/pbs_dataservice stop > /dev/null 2>&1
+		fi
+	else
+		for _svr in ${PBS_SERVER_INSTANCES//,/ }
```
Comment (subhasisb) [inline src/cmds/scripts/pbs_init.d.in:558]: Please provide a comment or call a separate function start/stop multi-servers, instead of such long functions

File: src/cmds/scripts/pbs_init.d.in [hunk 11]
```diff
@@ -417,96 +505,111 @@ start_pbs() {
...
+					kill ${_svr#*:}
+				done
+			fi
+			for _svr in ${_pids}
+			do
+				_port=${_svr%%:*}
+				_pid=${_svr#*:}
+				echo "PBS server (on port ${_port}) - was pid: ${_pid}"
+			done
+			${PBS_EXEC}/bin/qstat -Bf >/dev/null 2>&1
```
Comment (subhasisb) [inline src/cmds/scripts/pbs_init.d.in:585]: before we shut down database, are we sure that all the server instances are actually down?

File: src/cmds/scripts/pbs_init.d.in [hunk 12]
```diff
@@ -539,22 +642,35 @@ stop_pbs() {
     do
 	sleep 1
 	something_running=""
-	if [ "${PBS_START_SERVER}" -gt 0 ] ; then
-	    if check_prog "server" || check_prog "oldserver" ; then
-		something_running=" pbs_server"
-	    fi
+	if [ "${PBS_START_SERVER}" -gt 0 ]; then
```
Comment (subhasisb) [inline src/cmds/scripts/pbs_init.d.in:645]: why not make a function like stop_local_servers

File: src/cmds/scripts/pbs_habitat.in (hunk 8 above)
Comment (hirenvadalia) [reply src/cmds/scripts/pbs_habitat.in:538]: In this script, we need only one server, because in this we just create/initialize PBS DB schema/tables in DB, and then set few configs like license, create default node/queue, etc.
+ I am setting PBS_BATCH_SERVICE_PORT because without that server fails with wrong multiserver config error as it needs exact port number in get_server_index(self)

File: src/cmds/scripts/pbs_init.d.in (hunk 9 above)
Comment (hirenvadalia) [reply src/cmds/scripts/pbs_init.d.in:536]: Sure will check remove.

File: src/cmds/scripts/pbs_init.d.in (hunk 11 above)
Comment (hirenvadalia) [reply src/cmds/scripts/pbs_init.d.in:585]: Yes, We are using qstat -Bf to ensure that, as if even one server is running, qstat -Bf will succeed so will not stop DB. And if qstat -Bf fails means all servers are down so we stop DB also.

File: src/cmds/scripts/pbs_habitat.in (hunk 8 above)
Comment (subhasisb) [reply src/cmds/scripts/pbs_habitat.in:538]: but if not port is specified it should default to 15001

File: src/cmds/scripts/pbs_init.d.in (hunk 11 above)
Comment (subhasisb) [reply src/cmds/scripts/pbs_init.d.in]: ok

File: src/cmds/scripts/pbs_init.d.in [hunk 13]
```diff
@@ -582,66 +562,13 @@ stop_local_servers() {
 				_pid=${_svr#*:}
 				echo "PBS server (on port ${_port}) - was pid: ${_pid}"
 			done
-			${PBS_EXEC}/bin/qstat -Bf >/dev/null 2>&1
```
Comment (subhasisb) [inline src/cmds/scripts/pbs_init.d.in]: why did you remove the qstat -Bf? That was a good way to ensure that all servers are down

File: src/cmds/scripts/pbs_habitat.in (hunk 8 above)
Comment (hirenvadalia) [reply src/cmds/scripts/pbs_habitat.in:538]: Yes, true, and that is taken care in pbs_loadconf() right? so let's say `PBS_SERVER_INSTANCES=svr1:svr2:svr3` then here we do `export PBS_BATCH_SERVICE_PORT=` (because _first_svr_port will be empty), then in loadconf() it will default to 15001.

File: src/cmds/scripts/pbs_init.d.in (hunk 13 above)
Comment (hirenvadalia) [reply src/cmds/scripts/pbs_init.d.in]: I already part of pbs_dataservice script, in stop operation, pbs_dataservice script first checks for qstat -Bf, so here qstat check was redundant.

Comment (hirenvadalia) [reply src/cmds/scripts/pbs_init.d.in]: Here: https://github.com/hirenvadalia/pbspro/blob/ms_init/src/cmds/scripts/pbs_dataservice#L363
